frontend/src/metabase/query_builder/components/template_tags/SnippetForm/SnippetForm.tsx

```diff
@@ -0,0 +1,209 @@
+}
+
+interface SnippetFormDispatchProps {
+  handleCreateSnippet: (
+    snippet: SnippetFormValues,
+  ) => Promise<NativeQuerySnippet>;
+  handleUpdateSnippet: (
+    snippet: UpdateSnippetFormValues,
+  ) => Promise<NativeQuerySnippet>;
+}
+
+type SnippetFormProps = SnippetFormOwnProps &
+  SnippetLoaderProps &
+  SnippetFormDispatchProps;
+
+const mapDispatchToProps = {
+  handleCreateSnippet: Snippets.actions.create,
+  handleUpdateSnippet: Snippets.actions.update,
+};
+
+function SnippetForm({
+  snippet,
+  snippetCollections,
+  handleCreateSnippet,
+  handleUpdateSnippet,
+  onCreate,
+  onUpdate,
+  onArchive,
+  onCancel,
+}: SnippetFormProps) {
+  const isEditing = snippet.id != null;
+  const hasManyCollections = snippetCollections.length > 1;
+
+  const initialValues = useMemo(
+    () =>
+      SNIPPET_SCHEMA.cast(
+        {
+          ...snippet,
+          content: snippet.content || "",
+          parent_id: canonicalCollectionId(snippet.id),
+        },
+        { stripUnknown: true },
+      ),
+    [snippet],
+  );
+
+  const handleCreate = useCallback(
+    async (values: SnippetFormValues) => {
+      const action = await handleCreateSnippet(values);
+      const snippet = Snippets.HACK_getObjectFromAction(action);
+      onCreate?.(snippet);
+    },
+    [handleCreateSnippet, onCreate],
+  );
+
+  const handleUpdate = useCallback(
+    async (values: UpdateSnippetFormValues) => {
+      const action = await handleUpdateSnippet(values);
+      const nextSnippet = Snippets.HACK_getObjectFromAction(action);
+      onUpdate?.(nextSnippet, snippet as NativeQuerySnippet);
+    },
+    [snippet, handleUpdateSnippet, onUpdate],
+  );
+
+  const handleSubmit = useCallback(
+    async values => {
+      if (isEditing) {
+        await handleUpdate({ ...values, id: snippet.id });
+      } else {
+        await handleCreate(values);
+      }
+    },
+    [snippet.id, isEditing, handleCreate, handleUpdate],
+  );
+
+  const handleArchive = useCallback(async () => {
+    await handleUpdateSnippet({
+      id: snippet.id as NativeQuerySnippet["id"],
+      archived: true,
+    });
+    onArchive?.();
+  }, [snippet.id, handleUpdateSnippet, onArchive]);
+
+  return (
+    <FormProvider
+      initialValues={initialValues}
+      validationSchema={SNIPPET_SCHEMA}
+      onSubmit={handleSubmit}
+    >
+      {({ dirty }) => (
+        <Form disabled={!dirty}>
+          <FormSnippetTextArea
+            name="content"
+            title={t`Enter some SQL here so you can reuse it later`}
+            placeholder="AND canceled_at IS null\nAND account_type = 'PAID'"
+            autoFocus
+            rows={4}
+          />
+          <FormInput
+            name="name"
+            title={t`Give your snippet a name`}
+            placeholder={t`Current Customers`}
+          />
+          <FormInput
+            name="description"
+            title={t`Add a description`}
+            placeholder={t`It's optional but oh, so helpful`}
```

Comment:
`nullable`?

Reply:
Right, I always forget about this, thanks